Add 'Type transformation functions' section to order confirmation chapter (#10)
melange-re/melange-for-react-devs: docs/order-confirmation/index.md

> ## `Item.make`

Reviewer: I am not sure using these code-related titles will help. Maybe we could make them more human-friendly? Like "Rendering items" ?

Author: Renamed in https://github.com/melange-re/melange-for-react-devs/pull/10/commits/58da92ea8f4ed79fbf6d1a829715126d73f0ddce

Now we're ready to define the `Item.make` function which will render the `Item`
component:

> <<< Snippets.re#make

Reviewer: What are the benefits of keeping all snippets in a single file? Vs having a file for each module (`Item.re`, `Order.re`...)

Author: Split up Snippets into multiple modules: https://github.com/melange-re/melange-for-react-devs/pull/10/commits/08563e833f9a8def50c0ac73f41181c5c97fd84f

  [Js.Array.map](https://melange.re/v2.1.0/api/re/melange/Js/Array/index.html#val-map),
  which is the Melange binding to the [Array.map
  method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).
- There's a call to `React.array`, which we'll address in a little bit.

> ## `Index.re`

Reviewer: Same here, could be "Rendering the whole order"

Author: Renamed in https://github.com/melange-re/melange-for-react-devs/pull/10/commits/58da92ea8f4ed79fbf6d1a829715126d73f0ddce

Render the `Order` component inside `src/order-confirmation/Index.re`:

<<< Snippets.re#index

Run `make serve` inside `src/order-confirmation` to see your new app in action.

> ## `Js.Array.mapi`

Reviewer: Could be "Passing `key` prop to elements"

Author: Renamed in https://github.com/melange-re/melange-for-react-devs/pull/10/commits/58da92ea8f4ed79fbf6d1a829715126d73f0ddce

Open your browser's dev console, where you should see a warning:

```
Warning: Each child in a list should have a unique "key" prop.
```

> Oops, we forgot the set the `key` props! One way to fix this is to use

Reviewer: ```suggestion
Oops, we forgot the set the `key` prop! One way to fix this is to use
```

Author: Fixed in https://github.com/melange-re/melange-for-react-devs/pull/10/commits/58da92ea8f4ed79fbf6d1a829715126d73f0ddce

[Js.Array.mapi](https://melange.re/v2.1.0/api/re/melange/Js/Array/index.html#val-mapi)
instead[^4] so we can set `key` based on the index of the element:

<<< Snippets.re#mapi

The `Js.Array.mapi` function is also a binding to the `Array.map` method, but
unlike `Js.Array.map`, it passes the element *and the index* into the callback.
If you hover over it, you'll see that it has the type signature

```
(('a, int) => 'b, array('a)) => array('b)
```

When a JavaScript function has optional arguments, it's common to create
[multiple OCaml functions that bind to
it](https://melange.re/v2.1.0/communicate-with-javascript/#approach-1-multiple-external-functions).
We'll discuss this in more detail [later](/todo).

> ## `React.array`

Reviewer: "Making sure all elements have the same type"

Author: Renamed this section to "Type transformation functions" and added more detail along with an example
https://github.com/melange-re/melange-for-react-devs/pull/10/commits/9beeb2dca5208127ec9104366c790cc5aebaf531

       but an expression was expected of type React.element
```

> We get this error because the `tbody` element expects children of type

Reviewer: This is minor, but there's certain repetition between the "we'd get this error" in line 187 and the one in this line. Maybe change with

```suggestion
The compiler warns us because the `tbody` element expects children of type
```
? or st similar

Author: Rephrased in https://github.com/melange-re/melange-for-react-devs/pull/10/commits/f580d453198ae43a1718ead1b480af340c30bcbe
